spark/src/main/java/org/apache/iceberg/actions/PlanScanAction.java

```diff
@@ -0,0 +1,263 @@
+          scan.getClass()));
+    }
+  }
+
+  private DeleteFileIndex buildDeleteFileIndex() {
+    // Build delete index locally
+    List<ManifestFile> deleteManifests = snapshot.deleteManifests();
+    DeleteFileIndex.Builder deleteFileIndexBuilder = DeleteFileIndex.builderFor(table.io(), deleteManifests);
+    deleteFileIndexBuilder.caseSensitive(scan.isCaseSensitive());
+    deleteFileIndexBuilder.specsById(table.specs());
+    deleteFileIndexBuilder.filterData(scan.filter());
+    return deleteFileIndexBuilder.build();
+  }
+
+  private CloseableIterable<FileScanTask> planDataTableScanFiles() {
+    // TODO Currently this approach reads all manifests, no manifest filtering - Maybe through pushdowns or table read
+    // options?
+
+
+    // Create a dataframe of all DataFile entries
+    String dataFilesMetadataTable = metadataTableName(MetadataTableType.ENTRIES);
+    Dataset<Row> manifestEntries =
+        spark.read()
+            .format("iceberg")
+            .load(dataFilesMetadataTable);
+
+    // Todo pushdown filters to ManifestEntriesTable
+    // Read entries which are not deleted and are datafiles and not delete files
+    Dataset<Row> dataFileEntries = manifestEntries
+        .filter(manifestEntries.col("data_file").getField(DataFile.CONTENT.name()).equalTo(0)) // Only DataFiles
+        .filter(manifestEntries.col("status").notEqual(2)); // No Deleted Files
+
+    // Handle Incremental Scans
+    if (scan instanceof IncrementalDataTableScan) {
+      IncrementalDataTableScan incScan = (IncrementalDataTableScan) scan;
+      List<Snapshot> snapshots = IncrementalDataTableScan.snapshotsWithin(table, incScan.fromSnapshotId(),
+          incScan.toSnapshotId());
+      List<Long> validSnapshotIds = snapshots.stream().map(Snapshot::snapshotId).collect(Collectors.toList());
+      dataFileEntries = dataFileEntries
+          .filter(dataFileEntries.col("snapshot_id").isin(validSnapshotIds.toArray()));
+    }
+
+    // Build up evaluators and filters for Metrics and Partition values
+    final Expression scanFilter = scan.filter();
+    final boolean ignoreResiduals = scan.shouldIgnoreResiduals();
+    final boolean isCaseSensitive = scan.isCaseSensitive();
+
+    Map<Integer, PartitionSpec> specsById = scan.table().specs();
+    JavaSparkContext jsc = JavaSparkContext.fromSparkContext(spark.sparkContext());
+
+    // Build cache of partition evaluators
+    ImmutableMap.Builder<Integer, Evaluator> evalMapBuilder = ImmutableMap.builder();
+    specsById.entrySet().stream().forEach(entry ->
+        evalMapBuilder.put(entry.getKey(),
+            new Evaluator(entry.getValue().partitionType(),
+                Projections.inclusive(entry.getValue(), isCaseSensitive).project(scanFilter))));
+
+    Map<Integer, Evaluator> partitionEvaluatorsById = evalMapBuilder.build();
+    Broadcast<Map<Integer, Evaluator>> broadcastPartitionEvaluators = jsc.broadcast(partitionEvaluatorsById);
```

Comment:
Building this broadcast should probably be in a separate private method.

Reply:
Yeah, I'm pushing the complexity bounds pretty hard here. I'll move it out like the deletefileindex building and friends